oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsSoundCard.java

```diff
@@ -0,0 +1,127 @@
+ * https://github.com/oshi/oshi/graphs/contributors
+ */
+package oshi.hardware.platform.windows;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import oshi.hardware.common.AbstractSoundCard;
+import oshi.jna.platform.windows.WbemcliUtil;
+import oshi.util.platform.windows.WmiUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Sound Card data obtained by Wmi Queries for Windows.
+ *
+ * @author : BilalAM
+ */
+public class WindowsSoundCard extends AbstractSoundCard {
+
+    enum SoundCardName {
+        MANUFACTURER, NAME;
+    }
+
+    enum SoundCardKernel {
+        DRIVERPROVIDERNAME, DRIVERNAME, DRIVERVERSION, DEVICENAME;
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(WindowsSoundCard.class);
+    private static WbemcliUtil.WmiQuery<SoundCardName> AUDIO_CARD_QUERY = new WbemcliUtil.WmiQuery<>("Win32_SoundDevice",SoundCardName.class);
+    private static WbemcliUtil.WmiResult<SoundCardName> AUDIO_CARD_QUERY_RESULT = WmiUtil.queryWMI(AUDIO_CARD_QUERY);
+
+    public WindowsSoundCard(String kernelVersion, String name, String codec) {
+        super(kernelVersion, name, codec);
+    }
+
+    /**
+     * Gets a 'complete' name of the sound card. The 'complete' name is in fact a concatenated string from the Win32_SoundDevice query
+     * consisting of :
+     * <ul>
+     *     <li>The Manufacturer property</li>
+     *     <li>The Name property</li>
+     * </ul>
+     * @param index
+     *            The row to fetch from
+     * @return The audio card name
+     */
+    private static String getAudioCardCompleteName(int index) {
+        return WmiUtil.getString(AUDIO_CARD_QUERY_RESULT,SoundCardName.MANUFACTURER,index) + " " +
+                WmiUtil.getString(AUDIO_CARD_QUERY_RESULT,SoundCardName.NAME,index);
+    }
+
+    /**
+     * Gets the standard audio card name returned by the Name property of Win32_SoundDevice query.
+     * @param index
+     *          The row to fetch from.
+     * @return  The audio card name.
+     */
+    private static String getAudioCardName(int index) {
+        return WmiUtil.getString(AUDIO_CARD_QUERY_RESULT,SoundCardName.NAME,index);
+    }
+
+    /**
+     * Gets the 'complete' driver name . The 'complete' driver name is infact a concatenated string
+     * of the values returned by the AUDIO_CARD_KERNEL_QUERY query result consisting of :
+     * <ul>
+     *     <li>The DriverName property </li>
+     *     <li>The DriverVersion property</li>
+     *     <li>The DeviceName property</li>
+     *     <li>The DriverProvider property</li>
+     * </ul>
+     * Note: The query of AUDIO_CARD_KERNEL_QUERY is like Get-WmiObject Win32_PnPSignedDriver |
+     * where {$_.devicename -like "*(<b>the name of the card returned by Win32_SoundDevice Name Property</b>)*"}
+     * @param index
+     *        The row to fetch from
+     * @return  The 'complete' name of the driver.
+     */
+    private static String getAudioCardKernelVersion(int index) {
```

Comment:
This seems inefficient: you run one query to get the card names, and then run this query for each card using a filter.

I think it would be better to run this query first without a filter; iterate over the results and populate a Map using the DeviceName as the key and calculated string below as the Value.   Then iterate over the `Win32_SoundDevice` results and lookup the kernel version from the map using the same name key.

Reply:
The reason I run this with a filter is because Win32_PnPSignedDriver returns a very long list of information of all the drivers , instead of getting all the results and then populating a map , I made a filter query that 
- returns only the driver info about a particular device(in our case it is the sound card returned by `AUDIO_CARD_QUERY_RESULT` i.e Win32_SoundDevice) 
- Then builds a string consisting of the attributes returned by our query

Reply:
As far as my understanding the number of times Wmi gets queried 

- Once for Win32_SoundDevice.
- And depending on the number of cards returned by SoundDevice ,'n' times for Win32_PnPSignedDriver where n are the number of cards returned by Win32_SoundDevice.
So according to my understanding Wmi gets queried on a maximum of 2-3 times (if we consider a machine having 2 sound cards or more than that (rare) ) .
I may be wrong on this but this what I understand !

Reply:
You still end up iterating the SoundDevice query and doing multiple individual queries.

Take a look at `WindowsUsbDevice.java` and see how `PNPENTITY_BASE_CLASS` is defined and used.  We collect the list of PnPDeviceIDs and then do a single query to get just those device IDs.

So my suggestion:
(One time on initialization) 
1. Run the audio_card_query and store results in a MANUFACTURER_MAP with the name as the key and manufacturer as the value. 
2. Iterate the manufacturer map keys and build the kernel query where clause similar to the USB Device example I gave and run that query, storing the results in a KERNEL_VERSION_MAP with name as the key, and constructing the string as you currently do.
(When the getter is called)
3. Iterate the manufacturer map keys and build the sound card objects using the values from the two different maps.

Reply:
In the `WindowsUsbDevice.java` example , what you are trying to say is that I should make up my Win32_PnPSignedDriver query only once and then run it ?

- Like the query should be something like this 
`Win32_PnPSignedDriver WHERE DeviceName = "{name value from MANUFACTURER map}" OR "{second name value from MANUFACTURER map}"` and so on..till the size of the map , just like you have built it in the UsbClass.**And that I should do this only once in a static block.**
- Then when I call the `getSoundCards()` method , it is going to get the kernels by running the newly built query **ONLY ONCE** , then iterate over the results and build the sound cards accordingly ?

I hope I have understood this.

Reply:
Yes!

1. The list of sound cards doesn't ever change so you can pull that once on initialization, store in the map, and construct a single query like your first bullet.
2. When you call `getSoundCards()` you will run the saved query once (it will retrieve multiple rows) and iterate over the resultcounts, matching the name to the name/manufacturer map and building the sound card objects.
3. Repeat 2 as needed if information changes.

Reply:
Okay now I am getting it , will start working on this ASAP today.
Sorry for taking a bit long though..